library/analytics/src/main/java/org/cru/godtools/analytics/firebase/FirebaseAnalyticsService.kt

```diff
@@ -86,15 +110,58 @@ class FirebaseAnalyticsService @VisibleForTesting internal constructor(
         firebase.logEvent(event.firebaseEventName, bundle)
     }
 
+    // region Visitor ids
+    @WorkerThread
+    private fun MutableMap<String, Any?>.baseContextData(userProfile: UserInfo?, event: AnalyticsBaseEvent? = null) {
+        put(USER_PROP_APP_NAME, VALUE_GODTOOLS)
+        put(USER_PROP_CONTENT_LANGUAGE, LocaleCompat.toLanguageTag(Locale.getDefault()))
+        put(USER_PROP_LOGGED_IN_STATUS, "false")
+        if (userProfile != null) {
+            put(USER_PROP_LOGGED_IN_STATUS, "true")
+            put(USER_PROP_SSO_GUID, userProfile.ssoGuid)
+            userProfile.grMasterPersonId?.let { put(USER_PROP_GR_MASTER_PERSON_ID, it) }
+        }
+    }
+
+    private val userProfileStateFlow = oktaUserProfileProvider.userInfoFlow(refreshIfStale = false)
+        .onEach { analyticsExecutor.execute { updateVisitorIdIdentifiers(it) } }
+        .stateIn(coroutineScope, SharingStarted.Eagerly, null)
+
+    @WorkerThread
+    private fun updateVisitorIdIdentifiers(userProfile: UserInfo?) {
+        Visitor.syncIdentifiers(
+            mutableMapOf(
+                VISITOR_ID_GUID to userProfile?.ssoGuid,
+                VISITOR_ID_MASTER_PERSON_ID to userProfile?.grMasterPersonId
+            ), when {
+            userProfile != null -> VisitorID.VisitorIDAuthenticationState.VISITOR_ID_AUTHENTICATION_STATE_AUTHENTICATED
+            else -> VisitorID.VisitorIDAuthenticationState.VISITOR_ID_AUTHENTICATION_STATE_LOGGED_OUT
+        }
+        )
+    }
+    // endregion Visitor ids
+
     init {
         oktaUserProfileProvider.userInfoFlow()
             .map { it?.ssoGuid }
             .distinctUntilChanged()
-            .onEach { firebase.setUserId(it); firebase.setUserProperty("debug", BuildConfig.DEBUG.toString()) }
+            .onEach { it ->
+                firebase.setUserId(it)
+                analyticsExecutor.execute {
+                    updateVisitorIdIdentifiers(userProfileStateFlow.value)
+                    val data = buildMap {
+                        baseContextData(userProfileStateFlow.value)
```

Comment:
you actually don't need `userProfileStateFlow` here because you already have the current `UserInfo` object in `it`

Reply:
err, scratch that. you will need to rework a little bit of the previous 2 flow transforms so that you have the user profile in the `onEach`

Reply:
I would drop the `distinctUntilChanged` call, attributes besides the guid could change without the guid being updated. so you shouldn't be preventing updating user property settings if the guid hasn't changed but another attribute may have